apps/sms/services/js/moz_mobile_message/messaging_service.js

```diff
@@ -0,0 +1,191 @@
+/*global BridgeServiceMixin, Settings */
+'use strict';
+
+if (!('BridgeServiceMixin' in self)) {
+  importScripts('/services/js/bridge_service_mixin.js');
+}
+
+(function(exports) {
+
+const SERVICE_NAME = 'messaging-service';
+const MOBILE_MESSAGE_CLIENT_NAME = 'mozMobileMessageShim';
+
+const METHODS = Object.freeze([
+  'sendSMS', 'sendMMS', 'resendMessage'
+]);
+
+function ensureBridge() {
+  if (!('bridge' in self)) {
+    importScripts('/lib/bridge.js');
+  }
+}
+
+// 0 is a valid value so we need to take care to not consider it as a falsy
+// value. We want to return null for anything that's not a number or a
+// string containing a number.
+function sanitizeServiceId(serviceId) {
+  if (serviceId == null || // null or undefined
+      isNaN(+serviceId)) {
+    serviceId = null;
+  } else {
+    serviceId = +serviceId;
+  }
+
+  return serviceId;
+}
+
+function getSendOptionsFromServiceId(serviceId) {
+  var sendOpts;
+
+  if (serviceId != null && // not null, not undefined
+      Settings.hasSeveralSim()) {
+    sendOpts = { serviceId: serviceId };
+  }
+
+  return sendOpts;
+}
+
+var MessagingService = {
+  /**
+   * Initializes our service using the Service mixin's initService and
+   * requests mozMobileMessage shim as client.
+   */
+  init() {
+    ensureBridge();
+
+    this.initService();
+
+    this.mozMobileMessageClient = bridge.client(MOBILE_MESSAGE_CLIENT_NAME);
```

Comment:
My main reason is just like @julienw said, since the property could not be exposed if it's not defined in service, that's way I didn't make it into real "private" in the end. But still thanks for the suggestion because it should be essential pattern at least for shared component for view.

For the iframe endpoint, yeah I agree that it would be a problem if `endpoint` is necessary for returning the client. Implementing this logic might be better in my POV because service -> shim(iframe) communication is quite common currently. Making it into plugin is a good suggestion sounds a good idea since not all the service will need this, and hopefully we could get rid of the shim in the future(not the near future anyway...).